File: src/main/kotlin/com/planner/weekly/api/v1/NotificationController.kt
package com.planner.weekly.api.v1

import com.planner.weekly.domain.notification.NotificationService
import com.planner.weekly.dto.response.ApiResponse
import com.planner.weekly.dto.response.NotificationResponse
import com.planner.weekly.security.UserPrincipal
import io.swagger.v3.oas.annotations.Operation
import io.swagger.v3.oas.annotations.security.SecurityRequirement
import io.swagger.v3.oas.annotations.tags.Tag
import org.springframework.security.core.annotation.AuthenticationPrincipal
import org.springframework.web.bind.annotation.*

@RestController
@RequestMapping("/api/v1/notifications")
@Tag(name = "Notifications", description = "Notification APIs")
@SecurityRequirement(name = "bearerAuth")
class NotificationController(
    private val notificationService: NotificationService
) {

    @GetMapping
    @Operation(summary = "Get all notifications")
    fun getNotifications(
        @AuthenticationPrincipal user: UserPrincipal
    ): ApiResponse<List<NotificationResponse>> {
        val notifications = notificationService.getUserNotifications(user.id)
        return ApiResponse.ok(notifications.map { NotificationResponse.from(it) })
    }

    @GetMapping("/unread")
    @Operation(summary = "Get unread notifications")
    fun getUnreadNotifications(
        @AuthenticationPrincipal user: UserPrincipal
    ): ApiResponse<List<NotificationResponse>> {
        val notifications = notificationService.getUnreadNotifications(user.id)
        return ApiResponse.ok(notifications.map { NotificationResponse.from(it) })
    }

    @GetMapping("/unread/count")
    @Operation(summary = "Get unread notification count")
    fun getUnreadCount(
        @AuthenticationPrincipal user: UserPrincipal
    ): ApiResponse<Long> {
        val count = notificationService.getUnreadCount(user.id)
        return ApiResponse.ok(count)
    }

    @PostMapping("/{notificationId}/read")
    @Operation(summary = "Mark a notification as read")
    fun markAsRead(
        @AuthenticationPrincipal user: UserPrincipal,
        @PathVariable notificationId: String
    ): ApiResponse<NotificationResponse> {
        val notification = notificationService.markAsRead(notificationId, user.id)
        return ApiResponse.ok(NotificationResponse.from(notification))
    }

    @PostMapping("/read-all")
    @Operation(summary = "Mark all notifications as read")
    fun markAllAsRead(
        @AuthenticationPrincipal user: UserPrincipal
    ): ApiResponse<Unit> {
        notificationService.markAllAsRead(user.id)
        return ApiResponse.ok(Unit)
    }
}

File: src/notification/notification.controller.ts
import { Controller, Get, Post, Param } from '@nestjs/common';
import { ApiTags, ApiOperation, ApiBearerAuth } from '@nestjs/swagger';
import { NotificationService } from './notification.service';
import { NotificationResponseDto, UnreadCountResponseDto } from './dto/notification.dto';
import { CurrentUser } from '../common/decorators/current-user.decorator';
import { JwtPayload } from '../common/interfaces/request-with-user.interface';
import { ApiResponse } from '../common/dto/api-response.dto';

@ApiTags('Notifications')
@ApiBearerAuth()
@Controller('notifications')
export class NotificationController {
  constructor(private notificationService: NotificationService) {}

  @Get()
  @ApiOperation({ summary: '전체 알림 목록' })
  async getAllNotifications(
    @CurrentUser() user: JwtPayload,
  ): Promise<ApiResponse<NotificationResponseDto[]>> {
    const notifications = await this.notificationService.findAllByUser(user.sub);
    return ApiResponse.ok(notifications);
  }

  @Get('unread')
  @ApiOperation({ summary: '읽지 않은 알림' })
  async getUnreadNotifications(
    @CurrentUser() user: JwtPayload,
  ): Promise<ApiResponse<NotificationResponseDto[]>> {
    const notifications = await this.notificationService.findUnreadByUser(user.sub);
    return ApiResponse.ok(notifications);
  }

  @Get('unread/count')
  @ApiOperation({ summary: '읽지 않은 알림 수' })
  async getUnreadCount(
    @CurrentUser() user: JwtPayload,
  ): Promise<ApiResponse<UnreadCountResponseDto>> {
    const count = await this.notificationService.getUnreadCount(user.sub);
    return ApiResponse.ok({ count });
  }

  @Post(':notificationId/read')
  @ApiOperation({ summary: '알림 읽음 처리' })
  async markAsRead(
    @CurrentUser() user: JwtPayload,
    @Param('notificationId') notificationId: string,
  ): Promise<ApiResponse<NotificationResponseDto>> {
    const notification = await this.notificationService.markAsRead(
      notificationId,
      user.sub,
    );
    return ApiResponse.ok(notification);
  }

  @Post('read-all')
  @ApiOperation({ summary: '전체 알림 읽음 처리' })
  async markAllAsRead(
    @CurrentUser() user: JwtPayload,
  ): Promise<ApiResponse<null>> {
    await this.notificationService.markAllAsRead(user.sub);
    return ApiResponse.ok(null);
  }
}
